### ISSUE TITLE
Inline CSS broken when starting with a :root rule

### ISSUE THREAD

**Issue by fabiomrbarbosa** [11ty/eleventy#1201]
**Describe the bug**
My current project makes extensive use of CSS custom properties, which as per common practice I declare inside a `:root` selector. I'm also using SASS to process and minify my .scss stylesheets into one file, which I then  `{% include %}` via a Nunjucks template and filter as `safe` on the `<head>` of my website, to render it inline.

When I do this AND have an HTML entity declared somewhere else on my stylesheet (inside of a "content" property), Eleventy seems to automatically add some sort of unknown character right before `:root `, which ultimately means the browser skips all the variables I've declared inside of it.

If I add any regular selector before `:root`, my CSS works as intended.
If I manually copy and paste the minified CSS into the `<head>` of the website, it also works as intended.

**To Reproduce**
1. Write any CSS file starting with a `:root` selector and including an HTML entity (say, `content: "\2197";`);
2. Have it inlined via a Nunjucks `{% include %}` (haven't tried other template languages);
3. Filter the include as safe, or don't, the stylesheet breaks regardless;
4. Only part of styles are rendered, without any of the properties declared under `:root`.

**Expected behavior**
My inline CSS should be rendered just as I've coded it, without having to paste it manually in my template.

**Environment:**

- OS and Version: MacOS Mojave, Node v13.10.1, Sass 1.26.5
- Eleventy Version 0.11.0
- Build tasks: 
    "sass:process": "sass [source SCSS file] [output CSS file] --style=compressed",
    "build": "npm run sass:process && eleventy"

**Comment by Ryuno-Ki** [comment]
This rings a bell … I believe, I ran into something like this before. I assume, it's not Eleventy's fault.

Have you checked, what happens, if you include the compressed CSS via `<link />`?

Is it possible to attach a minimal CSS created this way?

**Comment by fabiobarbosa** [comment]
> This rings a bell … I believe, I ran into something like this before. I assume, it's not Eleventy's fault.
> 
> Have you checked, what happens, if you include the compressed CSS via `<link />`?
> 
> Is it possible to attach a minimal CSS created this way?

Right, maybe this is just a Nunjucks issue?
I'm attaching a minified CSS file. 

And indeed, everything does work properly if I include the compressed CSS via `<link />`.


[inline.min.css.txt](https://github.com/11ty/eleventy/files/4664217/inline.min.css.txt)

**Comment by Ryuno-Ki** [comment]
Hm, nunjucks is fine:

```js
const nunjucks = require('nunjucks')
const debug = nunjucks.renderString('{% include "inline.min.css" %}')
console.log(debug.charAt(0))
console.log(debug)
```

Looking into eleventy next.

```njk
<!doctype html>
<html>
<head>
<style>
{% include "inline.min.css" %}
</style>
</head>
<body>
</body>
```

and moved `inline.min.css` into an `_includes` directory, then running `node_modules/.bin/eleventy`.

Result:

```html
<!doctype html>
<html>
<head>
<style>
:root{--color-red: #c5004a;--color-darkred: #7f0036;--color-lightgray: #e0e0e0;--color-gray: #c0c0c0;--color-darkgray: #333;--color-navy: #17050f;--color-blue: #082840;--color-white: #fff;--fluid-min-width: 320;--fluid-max-width: 1440;--fluid-min-size: 16;--fluid-max-size: 21;--fluid-min-ratio: 1.2;--fluid-max-ratio: 1.25;--fluid-screen: 100vw;--fluid-bp: calc( (var(--fluid-screen) - ((var(--fluid-min-width) / 16) * 1rem)) / ((var(--fluid-max-width) / 16) - (var(--fluid-min-width) / 16)) );--fluid-max-negative: (1 / var(--fluid-max-ratio) / var(--fluid-max-ratio));--fluid-min-negative: (1 / var(--fluid-min-ratio) / var(--fluid-min-ratio));--fluid-min-scale--2: var(--fluid-min-scale--1) * var(--fluid-min-ratio) * var(--fluid-min-negative);--fluid-max-scale--2: var(--fluid-max-scale--1) * var(--fluid-max-ratio) * var(--fluid-max-negative);--fluid-min-size--2: (var(--fluid-min-size) * var(--fluid-min-scale--2)) / 16;--fluid-max-size--2: (var(--fluid-max-size) * var(--fluid-max-scale--2)) / 16;--size-step--2: calc( ( (var(--fluid-min-size--2) * 1rem) + (var(--fluid-max-size--2) - var(--fluid-min-size--2)) * var(--fluid-bp) ) );--fluid-min-scale--1: var(--fluid-min-ratio) * var(--fluid-min-negative);--fluid-max-scale--1: var(--fluid-max-ratio) * var(--fluid-max-negative);--fluid-min-size--1: (var(--fluid-min-size) * var(--fluid-min-scale--1)) / 16;--fluid-max-size--1: (var(--fluid-max-size) * var(--fluid-max-scale--1)) / 16;--size-step--1: calc( ( (var(--fluid-min-size--1) * 1rem) + (var(--fluid-max-size--1) - var(--fluid-min-size--1)) * var(--fluid-bp) ) );--fluid-min-scale-0: var(--fluid-min-ratio);--fluid-max-scale-0: var(--fluid-max-ratio);--fluid-min-size-0: (var(--fluid-min-size)) / 16;--fluid-max-size-0: (var(--fluid-max-size)) / 16;--size-step-0: calc( ( (var(--fluid-min-size-0) * 1rem) + (var(--fluid-max-size-0) - var(--fluid-min-size-0)) * var(--fluid-bp) ) );--fluid-min-scale-1: var(--fluid-min-scale-0) * var(--fluid-min-ratio);--fluid-max-scale-1: var(--fluid-max-scale-0) * var(--fluid-max-ratio);--fluid-min-size-1: (var(--fluid-min-size) * var(--fluid-min-scale-0)) / 16;--fluid-max-size-1: (var(--fluid-max-size) * var(--fluid-max-scale-0)) / 16;--size-step-1: calc( ( (var(--fluid-min-size-1) * 1rem) + (var(--fluid-max-size-1) - var(--fluid-min-size-1)) * var(--fluid-bp) ) );--fluid-min-scale-2: var(--fluid-min-scale-1) * var(--fluid-min-ratio);--fluid-max-scale-2: var(--fluid-max-scale-1) * var(--fluid-max-ratio);--fluid-min-size-2: (var(--fluid-min-size) * var(--fluid-min-scale-1)) / 16;--fluid-max-size-2: (var(--fluid-max-size) * var(--fluid-max-scale-1)) / 16;--size-step-2: calc( ( (var(--fluid-min-size-2) * 1rem) + (var(--fluid-max-size-2) - var(--fluid-min-size-2)) * var(--fluid-bp) ) );}@media screen and (min-width: 1440px){:root{--fluid-screen: calc(var(--fluid-max-width) * 1px)}}.entry-body{padding-left:var(--size-step-2);padding-right:var(--size-step-2)}.entry-body:not(:last-child)>div{border-bottom:1px solid var(--color-lightgray)}@media screen and (min-width: 56.25rem){.entry-body{padding-left:var(--size-step-4);padding-right:var(--size-step-4)}}.entry-content{margin:0 auto;max-width:var(--fluid-screen);padding-top:var(--size-step-2);padding-bottom:var(--size-step-2)}@media screen and (min-width: 56.25rem){.services__list{display:grid;grid-template-columns:1fr 2fr;grid-column-gap:var(--size-step-4)}}@font-face{font-family:"Inter";src:url("/assets/fonts/Inter.var.woff2") format("woff2");font-style:normal;font-display:swap;font-weight:100 900;unicode-range:U+0000-00FF,U+0131,U+0152-0153,U+02BB-02BC,U+02C6,U+02DA,U+02DC,U+2000-206F,U+2074,U+20AC,U+2122,U+2191,U+2193,U+2212,U+2215,U+FEFF,U+FFFD}html,body{padding:0;margin:0;font-family:"Inter",system-ui,sans-serif;font-size:var(--size-step-0);line-height:var(--size-step-2);letter-spacing:-0.02rem;color:var(--color-darkgray);background-color:var(--color-white);box-sizing:border-box}*,*:before,*:after{box-sizing:inherit}h2,h3{font-size:var(--size-step-0);line-height:var(--size-step-2);margin-top:0;margin-bottom:var(--size-step-2)}p,ol,ul{line-height:var(--size-step-2);margin-top:0;margin-bottom:var(--size-step-2)}p:last-child,ol:last-child,ul:last-child{margin-bottom:0}p,img{max-width:37.5rem}a{color:var(--color-blue)}a:visited{color:var(--color-navy)}@media screen and (min-width: 56.25rem){.entry-content>*:not(div){margin-left:calc(33.3% + var(--size-step-2))}}.services__list:first-of-type{padding-top:var(--size-step-2);border-top:1px solid var(--color-lightgray)}.services__list:not(:last-child){margin-bottom:var(--size-step-2);padding-bottom:var(--size-step-2);border-bottom:1px solid var(--color-lightgray)}.services__list ul{list-style:none;padding-left:0;columns:2}/*# sourceMappingURL=main.min.css.map */
</style>
</head>
<body>
</body>
```

Renders fine.

**Comment by fabiobarbosa** [comment]
Thanks for trying. In that case I've got no idea how to troubleshoot this further, except maybe by mentioning I forgot to add a pseudo-class `:after` selector with an HTML entity in the `content`, as I said in the beginning — like `content: "\2197";`, for instance.

I guess I'll just change the order of my imports on SCSS to have the reset appear before the :root properties or something, until I find a better solution.

**Comment by denisbrodbeck** [comment]
> like `content: "\2197";`, for instance.

Is `content: '\2197'` a valid unicode specifier? I always assumed that the syntax for unicode code points was `\u2197`.

Another possible issue could be that your template (where you're inserting the css inline) has somewhere an invisible control character. Best way to resolve this kind of issue is to delete the whole line and to type in your include logic by hand.

Maybe something upstream is generating the CSS in `utf-8` with a `BOM` before the actual content? This could be trimmed by using something like [strip-bom](https://www.npmjs.com/package/strip-bom), but best way is to remove the BOM in the source files.

**Comment by fabiomrbarbosa** [comment]
Thanks for that tip. However I've been checking the source files and none seem to include any BOM or control characters. 

Concerning the unicode specifier, it seems both `\2197` and `\u2197` behave exactly the same way, both with the bug reproducing scenario and with the workaround.

**Comment by fabiomrbarbosa** [comment]
I'm writing again to inform that in fact this situation seems to only happen on the local development server, while running `eleventy --serve`.

Indeed both the local minified CSS file and the production styles seem to be working, but when I inspect the source code from my local environment and copy and paste the inline CSS that is displayed there is an unrecognised character right in the beginning, which causes random breakage depending on the order of the elements (thus, not just unicode characters or CSS custom properties, which is something I've only been able to verify now that the project has expanded).

I guess I should be checking for Browsersync bugs?

**Comment by fabiomrbarbosa** [comment]
Ok, so I think I've got this down. Sorry in advance for the bother.

Indeed I just realised the minified CSS that my SASS installation is producing does come with a BOM before the content, as @denisbrodbeck mentioned. I've only noticed that after I realised that the production environment (Netlify in this case) was displaying the correct stylesheet but the local environment was not. 

Anyway, the issue seems to stem from the absence of a `@charset utf-8` from the main SCSS file, so whether it is a bug or not, it's definitely due to a lapse of mine.

Thank you all for your attention. I'll be closing this issue myself.
